man/kcutilcodec.1
.TH "KCUTILCODEC" 1 "2010-07-07" "Man Page" "Kyoto Cabinet"

.SH NAME
kcutilcodec \- a command

.SH DESCRIPTION
.PP
coming soon.

.SH SEE ALSO
.PP
.BR kcutiltest (1),
.BR kcprototest (1),
.BR kccachetest (1)


ChangeLog
2010-08-18  Mikio Hirabayashi  <info@fallabs.com>

	* kcplantdb.h (PlantDB::rcomp): new method.

	* kctreemgr.cc (runlist, proclist): descending order was added.

	- Release: 1.2.7

2010-08-17  Mikio Hirabayashi  <info@fallabs.com>

	* kcthread.cc (CondVar): Win32 support was added.

	* kcdb.h (jump_back, step_back): new functions.

	* kcplantdb.h (jump_last, step_back): new functions.

	* kcutiltest.cc (runpara, procpara): parallel tests were added.

	- Release: 1.2.6

2010-08-07  Mikio Hirabayashi  <info@fallabs.com>

	* kccachedb.h (CacheDB::clear_slot): a bug related to transaction was fixed.

	* kchashdb.h (HashDB::reorganize_file): a bug of overflow was fixed.

	- Release: 1.2.5

2010-08-06  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (hashpath): variance was improved.

	* kccompress.h (LZO::calculate_crc, LZMA::calculate_crc): new functions.

	- Release: 1.2.4

2010-08-01  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.cc (getsysinfo): some information was added for Darwin and Win32.

	* kcutil.cc (setstdiobin): new function.

	* kccompress.h (LZO, LZMA): new classes.

	* kcfile.cc (File::remove, File::rename): modified for Win32 virus checkers.

	* kcpolydb.h (PolyDB::open): "zcomp=lzo" and "zcomp=lzma" were added.

	- Release: 1.2.3

2010-07-29  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.cc (getsysinfo): new function.

	* kccache.h (GrassDB): new class.

	* kccachetest.cc: new file.

	- Release: 1.2.2

2010-07-28  Mikio Hirabayashi  <info@fallabs.com>

	* kcompress.h (ArcfourComressor::ArcfourComressor): suppressed a warning.

	* kccompress.h (ArcfourCompressor::set_compressor): new method.

	* kcpolydb.h (PolyDB::open): "zcomp=arcz" was added.

	* Makefile.in: now deals with an internal bug of GCC 4.2.1 on Mac OS X.

	- Release: 1.2.1

2010-07-26  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (BasicDB): renamed from FileDB.

	* kcplantdb.h: new file.

	* kctreedb.h: removed.

	* kchashdb.h (TreeDB): re-defined as a type alias.

	* kcdirdb.h (ForestDB): new class as a type alias.

	- Release: 1.2.0

2010-07-24  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (arccipher): new function.

	* kccopmress.h (ArcfourCompressor): new class.

	* kchashdb.h (HashDB::open): synchronize came to be called when auto sync mode.

	* kcdirdb.h (DirDB::commit_transaction): synchronize came to be called.

	* kcdirmgr.cc (dbmetaprint): status information became detail.

	* kcpolydb.h (PolyDB::open): "zcomp" and "zkey" were added.

	- Release: 1.1.1

2010-07-06  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (hexencode): performance was improved.

	* kcfile.cc (read_file, write_file): new functions.

	* kcdirdb.h, kcdirdb.cc, kcdirtest.cc, kcdirmgr.cc: new files.

	- Release: 1.1.0

2010-07-04  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (typestring): refactored.

	- Release: 1.0.4

2010-06-19  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h: useless virtual attributes of some functions were abolished.

	* kctreedb.h (synchronize): a bug related to a call back parameter was fixed.

	* kcpolytest.cc (procqueue): a bug related to a log message was fixed.

	- Release: 1.0.3

2010-06-05  Mikio Hirabayashi  <info@fallabs.com>

	* cmdcommon.h (mysrand): new function.

	* cmdcommon.h (myrand): algorithm was changed.

	- Release: 1.0.2

2010-05-25  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::open): error messages were added.

	* kcdb.h (DB::dump_snapshot, DB::load_snapshot): error codes were modified.

	* kchashdb.h (HashDB::open): error codes were modified.

	- Release: 1.0.1

2010-05-19  Mikio Hirabayashi  <info@fallabs.com>

	* kccommon.h: built-in macros with confliction were voided.

	* kclangc.cc (kcdbcopy, kccursetvalue): new functions.

	* kclangc.cc (kcatoi, kcatoix, kcatof): new functions.

	* kclang.cc (kcnan, kcinf, kcchknan, kcchkinf): new functions.

	- Release: 1.0.0

2010-05-06  Mikio Hirabayashi  <info@fallabs.com>

	* kccachetest.cc (procorder): a bug about record counting was fixed.

	- Release: 0.9.18

2010-05-02  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (FileProcessor::process): parameters for meta data was modified.

	- Release: 0.9.17

2010-04-28  Mikio Hirabayashi  <info@fallabs.com>

	* kccommon.h (_yield_, _testyield_, _assert_): new macros.

	* kcfile.h (File::get_current_directory, File::set_current_directory): new functions.

	- Release: 0.9.16

2010-04-21  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (inf, chknan, chkinf): new functions.

	* kcfile.h (File::rename, File::read_directory): new function.

	* kcdb.h (FileDB::copy): new function.

	* kccommon.h: FreeBSD and Solaris support was added.

	- Release: 0.9.15

2010-04-20  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::close): unlocking operation was added.

	* kcdb.h (FileProcessor::process): the offset parameter was abolished.

	- Release: 0.9.14

2010-04-15  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (nan): new function.

	* kcdb.h (File::DB::FileProcessor::process): parameters were added.

	* kclangc.cc (kchashmurmur, kchashfnv): new functions.

	* kcthread.cc: Win32 support was added.

	* kcfile.cc: Win32 support was added.

	* VCmakefile: new file.

	- Release: 0.9.13

2010-04-13  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (DB::typestring): a miscellaneous type was added.

	* kcpolydb.h (PolyDB::PolyDB): a constructor with a database object was added.

	* kcthread.cc (Mutex::lock_try): Mac OS X support was added.

	- Release: 0.9.12

2010-04-06  Mikio Hirabayashi  <info@fallabs.com>

	* kccommon.h: C++0x check was added.

	* kcdb.h (DB::dump_snapshot, DB::load_snapshot): new functions.

	* kcdb.h (DB::Cursor::set_value): new function.

	* kchashmgr.c (procdump, procload): new functions.

	* kctreemgr.c (procdump, procload): new functions.

	* kclangc.cc (kcmalloc, kcdbdumpsnap, kcdbloadsnap): new functions.

	- Release: 0.9.11

2010-04-01  Mikio Hirabayashi  <info@fallabs.com>


kcutilcodec.cc
/*************************************************************************************************
 * Popular encoders and decoders
 *                                                               Copyright (C) 2009-2010 FAL Labs
 * This file is part of Kyoto Cabinet.
 * This program is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version
 * 3 of the License, or any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *************************************************************************************************/


#include "cmdcommon.h"


// global variables
const char* g_progname;                  // program name


// function prototypes
int main(int argc, char** argv);
static void usage();
static int runhex(int argc, char** argv);
static int runciph(int argc, char** argv);
static int runcomp(int argc, char** argv);
static int runhash(int argc, char** argv);
static int runconf(int argc, char** argv);
static int32_t prochex(const char* file, bool dec);
static int32_t procciph(const char* file, const char* key);
static int32_t proccomp(const char* file, int32_t mode, bool dec);
static int32_t prochash(const char* file, int32_t mode);
static int32_t procconf(int32_t mode);


// main routine
int main(int argc, char** argv) {
  g_progname = argv[0];
  kc::setstdiobin();
  if (argc < 2) usage();
  int32_t rv = 0;
  if (!std::strcmp(argv[1], "hex")) {
    rv = runhex(argc, argv);
  } else if (!std::strcmp(argv[1], "ciph")) {
    rv = runciph(argc, argv);
  } else if (!std::strcmp(argv[1], "comp")) {
    rv = runcomp(argc, argv);
  } else if (!std::strcmp(argv[1], "hash")) {
    rv = runhash(argc, argv);
  } else if (!std::strcmp(argv[1], "conf")) {
    rv = runconf(argc, argv);
  } else if (!std::strcmp(argv[1], "version") || !std::strcmp(argv[1], "--version")) {
    printversion();
    rv = 0;
  } else {
    usage();
  }
  return rv;
}


// print the usage and exit
static void usage() {
  eprintf("%s: popular encoders and decoders of Kyoto Cabinet\n", g_progname);
  eprintf("\n");
  eprintf("usage:\n");
  eprintf("  %s hex [-d] [file]\n", g_progname);
  eprintf("  %s ciph [-key str] [file]\n", g_progname);
  eprintf("  %s comp [-def|-gz|-lzo|-lzma] [-d] [file]\n", g_progname);
  eprintf("  %s hash [-fnv|-path|-crc] [file]\n", g_progname);
  eprintf("  %s conf [-v|-i|-l|-p]\n", g_progname);
  eprintf("\n");
  std::exit(1);
}


// parse arguments of hex command
static int runhex(int argc, char** argv) {
  const char* file = NULL;
  bool dec = false;
  for (int32_t i = 2; i < argc; i++) {
    if (argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "-d")) {
        dec = true;
      } else {
        usage();
      }
    } else if (!file) {
      file = argv[i];
    } else {
      usage();
    }
  }
  int rv = prochex(file, dec);
  return rv;
}


// parse arguments of ciph command
static int runciph(int argc, char** argv) {
  const char* file = NULL;
  const char* key = "";
  for (int32_t i = 2; i < argc; i++) {
    if (argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "-key")) {
        if (++i >= argc) usage();
        key = argv[i];
      } else {
        usage();
      }
    } else if (!file) {
      file = argv[i];
    } else {
      usage();
    }
  }
  int rv = procciph(file, key);
  return rv;
}


// parse arguments of comp command
static int runcomp(int argc, char** argv) {
  const char* file = NULL;
  int32_t mode = 0;
  bool dec = false;
  for (int32_t i = 2; i < argc; i++) {
    if (argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "-def")) {
        mode = 1;
      } else if (!std::strcmp(argv[i], "-gz")) {
        mode = 2;
      } else if (!std::strcmp(argv[i], "-lzo")) {
        mode = 3;
      } else if (!std::strcmp(argv[i], "-lzma")) {
        mode = 4;
      } else if (!std::strcmp(argv[i], "-d")) {
        dec = true;
      } else {
        usage();
      }
    } else if (!file) {
      file = argv[i];
    } else {
      usage();
    }
  }
  int rv = proccomp(file, mode, dec);
  return rv;
}


// parse arguments of hash command
static int runhash(int argc, char** argv) {
  const char* file = NULL;
  int32_t mode = 0;
  for (int32_t i = 2; i < argc; i++) {
    if (argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "-fnv")) {
        mode = 1;
      } else if (!std::strcmp(argv[i], "-path")) {
        mode = 2;
      } else if (!std::strcmp(argv[i], "-crc")) {
        mode = 3;
      } else {
        usage();
      }
    } else if (!file) {
      file = argv[i];
    } else {
      usage();
    }
  }
  int rv = prochash(file, mode);
  return rv;
}


// parse arguments of conf command
static int runconf(int argc, char** argv) {
  int32_t mode = 0;
  for (int32_t i = 2; i < argc; i++) {
    if (argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "-v")) {
        mode = 'v';
      } else if (!std::strcmp(argv[i], "-i")) {
        mode = 'i';
      } else if (!std::strcmp(argv[i], "-l")) {
        mode = 'l';
      } else if (!std::strcmp(argv[i], "-p")) {
        mode = 'p';
      } else {
        usage();
      }
    } else {
      usage();
    }
  }
  int rv = procconf(mode);
  return rv;
}


// perform hex command
static int32_t prochex(const char* file, bool dec) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  if (dec) {
    char c;
    while (is->get(c)) {
      int32_t cc = (unsigned char)c;
      int32_t num = -1;
      if (cc >= '0' && cc <= '9') {
        num = cc - '0';
      } else if (cc >= 'a' && cc <= 'f') {
        num = cc - 'a' + 10;
      } else if (cc >= 'A' && cc <= 'F') {
        num = cc - 'A' + 10;
      }
      if (num >= 0) {
        if (is->get(c)) {
          cc = (unsigned char)c;
          if (cc >= '0' && cc <= '9') {
            num = num * 0x10 + cc - '0';
          } else if (cc >= 'a' && cc <= 'f') {
            num = num * 0x10 + cc - 'a' + 10;
          } else if (cc >= 'A' && cc <= 'F') {
            num = num * 0x10 + cc - 'A' + 10;
          }
          std::cout << (char)num;
        } else {
          std::cout << (char)num;
          break;
        }
      }
    }
    if (istr) std::cout << std::endl;
  } else {
    bool mid = false;
    char c;
    while (is->get(c)) {
      if (mid) std::cout << ' ';
      int32_t cc = (unsigned char)c;
      int32_t num = (cc >> 4);
      if (num < 10) {
        std::cout << (char)('0' + num);
      } else {
        std::cout << (char)('a' + num - 10);
      }
      num = (cc & 0x0f);
      if (num < 10) {
        std::cout << (char)('0' + num);
      } else {
        std::cout << (char)('a' + num - 10);
      }
      mid = true;
    }
    std::cout << std::endl;
  }
  return 0;
}


// perform ciph command
static int32_t procciph(const char* file, const char* key) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  char* cbuf = new char[ostr.size()];
  kc::arccipher(ostr.data(), ostr.size(), key, std::strlen(key), cbuf);
  std::cout.write(cbuf, ostr.size());
  delete[] cbuf;
  return 0;
}


// perform comp command
static int32_t proccomp(const char* file, int32_t mode, bool dec) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  bool err = false;
  switch (mode) {
    default: {
      kc::ZLIB::Mode zmode;
      switch (mode) {
        default: zmode = kc::ZLIB::RAW; break;
        case 1: zmode = kc::ZLIB::DEFLATE; break;
        case 2: zmode = kc::ZLIB::GZIP; break;
      }
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::ZLIB::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::ZLIB::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
    case 3: {
      kc::LZO::Mode zmode = kc::LZO::RAW;
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::LZO::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::LZO::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
    case 4: {
      kc::LZMA::Mode zmode = kc::LZMA::RAW;
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::LZMA::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::LZMA::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
  }
  return err ? 1 : 0;
}


// perform hash command
static int32_t prochash(const char* file, int32_t mode) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  switch (mode) {
    default: {
      uint64_t hash = kc::hashmurmur(ostr.data(), ostr.size());
      iprintf("%016llx\n", (unsigned long long)hash);
      break;
    }
    case 1: {
      uint64_t hash = kc::hashfnv(ostr.data(), ostr.size());
      iprintf("%016llx\n", (unsigned long long)hash);
      break;
    }
    case 2: {
      char name[kc::NUMBUFSIZ];
      uint32_t hash = kc::hashpath(ostr.data(), ostr.size(), name);
      iprintf("%s\t%08lx\n", name, (unsigned long)hash);
      break;
    }
    case 3: {
      uint32_t hash = kc::ZLIB::calculate_crc(ostr.data(), ostr.size());
      iprintf("%08x\n", (unsigned)hash);
      break;
    }
  }
  return 0;
}


// perform conf command
static int32_t procconf(int32_t mode) {
  switch (mode) {
    case 'v': {
      iprintf("%s\n", kc::VERSION);
      break;
    }
    case 'i': {
      iprintf("%s\n", _KC_APPINC);
      break;
    }
    case 'l': {
      iprintf("%s\n", _KC_APPLIBS);
      break;
    }
    case 'p': {
      iprintf("%s\n", _KC_BINDIR);
      break;
    }
    default: {
      iprintf("VERSION: %s\n", kc::VERSION);
      iprintf("LIBVER: %d\n", kc::LIBVER);
      iprintf("LIBREV: %d\n", kc::LIBREV);
      iprintf("FMTVER: %d\n", kc::FMTVER);
      iprintf("SYSNAME: %s\n", kc::SYSNAME);
      iprintf("BIGEND: %d\n", kc::BIGEND);
      iprintf("CLOCKTICK: %d\n", kc::CLOCKTICK);
      iprintf("PAGESIZE: %d\n", kc::PAGESIZE);
      iprintf("FEATURES: %s\n", kc::FEATURES);
      iprintf("TYPES: void*=%d short=%d int=%d long=%d long_long=%d size_t=%d"
              " float=%d double=%d long_double=%d\n",
              (int)sizeof(void*), (int)sizeof(short), (int)sizeof(int), (int)sizeof(long),
              (int)sizeof(long long), (int)sizeof(size_t),
              (int)sizeof(float), (int)sizeof(double), (int)sizeof(long double));
      std::map<std::string, std::string> info;
      kc::getsysinfo(&info);
      if (info["mem_total"].size() > 0)
        iprintf("MEMORY: total=%s free=%s cached=%s\n",
                info["mem_total"].c_str(), info["mem_free"].c_str(),
                info["mem_cached"].c_str());
      if (std::strcmp(_KC_PREFIX, "*")) {
        iprintf("prefix: %s\n", _KC_PREFIX);
        iprintf("includedir: %s\n", _KC_INCLUDEDIR);
        iprintf("libdir: %s\n", _KC_LIBDIR);
        iprintf("bindir: %s\n", _KC_BINDIR);
        iprintf("libexecdir: %s\n", _KC_LIBEXECDIR);
        iprintf("appinc: %s\n", _KC_APPINC);
        iprintf("applibs: %s\n", _KC_APPLIBS);
      }
      break;
    }
  }
  return 0;
}



// END OF FILE
